tests/unit/adapters/osf-adapter-test.js

```diff
@@ -1,12 +1,401 @@
+        });
+    });
+});
+
+test('#_deleteRelated defers to _doRelatedRequest, and unloads the deleted records', function(assert) {
+    let node = FactoryGuy.make('node', 'hasContributors');
+    let contrib = node.get('contributors').objectAt(1);
+    node.get('contributors').removeObject(contrib);
+
+    var unloadStub = this.stub(contrib, 'unloadRecord');
+    var doRelatedStub = this.stub(OsfAdapter.prototype, '_doRelatedRequest', () => {
+        return new Ember.RSVP.Promise(resolve => resolve());
+    });
+
+    Ember.run(() => {
+        node.save().then(() => {
+            assert.ok(doRelatedStub.calledOnce);
+            assert.ok(unloadStub.calledOnce);
+        }, () => {
+            // Fail
+            assert.ok(false);
+        });
+    });
+});
+
+test('#_doRelatedRequest with array', function(assert) {
+    let adapter = this.subject();
+
+    this.inject.service('store');
+    let store = this.store;
+
+    let node = FactoryGuy.make('node');
+    Ember.run.begin();
+    let children = FactoryGuy.buildList('node', 3).data.map(json => {
+        return store.createRecord('node', store.normalize('node', json).data.attributes);
+    });
+    Ember.run.end();
+
+    var mockAjax = this.stub(adapter, 'ajax', () => {
+        return new Ember.RSVP.Promise(resolve => resolve({}));
+    });
+    adapter._doRelatedRequest(
+        store,
+        node._internalModel.createSnapshot(),
+        children.map(c => c._internalModel.createSnapshot()),
+        'children',
+        'https://api.osf.io/v2/nodes/foobar/children/',
+        'POST'
+    );
+    var data = mockAjax.args[0][2].data.data;
+    assert.equal(data[0].attributes.title, children[0].get('title'));
+    assert.equal(data[1].attributes.title, children[1].get('title'));
+    assert.equal(data[2].attributes.title, children[2].get('title'));
 });
 
-// Replace this with your real tests.
-test('it exists', function(assert) {
-  let adapter = this.subject();
-  assert.ok(adapter);
+test('#_doRelatedRequest with single snapshot', function(assert) {
+    let adapter = this.subject();
+
+    this.inject.service('store');
+    let store = this.store;
+
+    let node = FactoryGuy.make('node');
+    Ember.run.begin();
+    let child = store.createRecord(
+        'node',
+        store.normalize('node', FactoryGuy.build('node').data).data.attributes
+    );
+    Ember.run.end();
+
+    var mockAjax = this.stub(adapter, 'ajax', () => {
+        return new Ember.RSVP.Promise(resolve => resolve({}));
+    });
+    adapter._doRelatedRequest(
+        store,
+        node._internalModel.createSnapshot(),
+        child._internalModel.createSnapshot(),
+        'children',
+        'https://api.osf.io/v2/nodes/foobar/children/',
+        'POST'
+    );
+    var data = mockAjax.args[0][2].data.data;
+    assert.equal(data.attributes.title, child.get('title'));
+});
+
+test('#_handleRelatedRequest makes correct calls for each change argument', function(assert) {
```

Comment:
Do we have a test to ensure that a record is correctly saved when we change both relationships *and* the original record? Eg if we add contributors and also change node title, we want to ensure that all changes get saved.

This is about the closest I could find (it checks that multiple operations are performed), but it doesn't seem to check anything about the original record.